### PR TITLE
Convert warning package to TypeScript

### PR DESCRIPTION
**Pull request by jpstevens**
## What?
This PR converts the `warning` package to TypeScript.

## Why?
Ensures package is fully type checked.

## How?
- converted files to `.ts`
- added explicit return types for functions in `index` and `utils` files

## Testing Instructions
`npm run test:unit -- packages/warning` tests pass
`npm run build:package-types` returns a zero exit code

### CODE REVIEW

**Comment by youknowriad** [comment]
Thanks for your help typing all these packages.

I thought this package was already fully typed. Is renaming to ts files necessary? 

@sirreal Can you help me navigate this a bit :P

**Comment by sirreal** [comment]
> I thought this package was already fully typed. Is renaming to ts files necessary?

Changing the file extension doesn't change much here. This package is typechecked, types are generated for the package, and no complex types are necessary that can't be easily expressed in JSDoc.

The type of the public API of the current version this package can be found here:
https://unpkg.com/browse/@wordpress/warning@2.57.0/build-types/index.d.ts

That would be exactly the same.

The explicit return type annotations added to the two functions might be a nice change, they're both inferred to be the same thing, but some would argue that adding those annotations is a good thing.

File: packages/warning/src/index.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ function isDev() {
  * }
  * ```
  */
-export default function warning( message ) {
+export default function warning( message: string ): void {
```
Comment (youknowriad) [inline packages/warning/src/index.ts:27]: I'm guessing that we need to remove the types from the JSDocs to be consistent with how we document ts files.

File: packages/warning/src/utils.ts [hunk 2]
```diff
@@ -4,4 +4,4 @@
  *
  * @type {Set<string>}
```
Comment (youknowriad) [inline packages/warning/src/utils.ts:5]: Should probably remove this as well right?

File: packages/warning/src/index.ts (hunk 1 above)
Comment (sirreal) [reply packages/warning/src/index.ts:27]: Yes. If we change the files to TypeScript we should definitely remove the types from JSDoc 💯

**Comment by youknowriad** [comment]
I've been thinking about this PR. While I see some value in it, I'm not sure that we should move forward with it. It has no real impact on the generated types or whether the package is fully typed. It's really a code style change.

And the thing is that while for this package it's fine, it may create a snowball effect for other packages and we'll find our selves spending time rewriting JS to TS for a lot of code and reviewing a lot of code for a small benefit.

**Comment by jsnajdr** [comment]
My main argument in favor of TS conversions like this is that the JSDoc type annotations are a rather obscure feature and most people are not familiar with them. I guess that many would be surprised to learn that such a thing even exists.

Who is really fluent in knowing that `@template {string} T` means `<T extends string>` or what's the difference between `@prop {string?}` and `@prop {string=}`? I believe the native syntax is more readable for virtually everyone, and is much better searchable in docs, Stack Overflow questions etc.

**Comment by youknowriad** [comment]
I don't feel that strongly, so I can be convinced, happy to hear other opinions.

**Comment by sirreal** [comment]
👋 @jpstevens If you'll update this PR on the latest `trunk` and remove types from JSDoc, we can likely give this a final review and land it.

**Pull request closed without merging**